pkg/grpc/marshaler_test.go

```diff
@@ -0,0 +1,86 @@
+package grpc
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/grpc/authz/allow"
+	"github.com/stretchr/testify/suite"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type MarshalerTest struct {
+	suite.Suite
+}
+
+func (a *MarshalerTest) SetupTest() {
+	// In order to start the gRPC server, we need to have the MTLS environment variables
+	// pointing to some valid certificate/key pair. In this case we are using the ones
+	// created for local-sensor, which are dummy self-signed certificates.
+	a.T().Setenv("ROX_MTLS_CERT_FILE", "../../tools/local-sensor/certs/cert.pem")
+	a.T().Setenv("ROX_MTLS_KEY_FILE", "../../tools/local-sensor/certs/key.pem")
+	a.T().Setenv("ROX_MTLS_CA_FILE", "../../tools/local-sensor/certs/caCert.pem")
+	a.T().Setenv("ROX_MTLS_CA_KEY_FILE", "../../tools/local-sensor/certs/caKey.pem")
+}
+func Test_MarshallerTest(t *testing.T) {
+	suite.Run(t, new(MarshalerTest))
+}
+
+var _ suite.SetupTestSuite = &MarshalerTest{}
+
+// Testing server error response from gRPC Gateway.
+type supressCveServiceTestErrorImpl struct {
+	v1.UnimplementedNodeCVEServiceServer
+}
+
+func (s *supressCveServiceTestErrorImpl) RegisterServiceServer(grpcServer *grpc.Server) {
+	v1.RegisterNodeCVEServiceServer(grpcServer, s)
+}
+
+func (s *supressCveServiceTestErrorImpl) RegisterServiceHandler(ctx context.Context, mux *runtime.ServeMux, grpcServer *grpc.ClientConn) error {
+	return v1.RegisterNodeCVEServiceHandler(ctx, mux, grpcServer)
+}
+
+func (s *supressCveServiceTestErrorImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, allow.Anonymous().Authorized(ctx, fullMethodName)
+}
+
+func (s *supressCveServiceTestErrorImpl) SuppressCVEs(_ context.Context, req *v1.SuppressCVERequest) (*v1.Empty, error) {
+	return nil, status.Error(codes.Canceled, req.String())
+}
+
+func (a *MarshalerTest) TestDurationParsing() {
+	url := "https://localhost:8080/v1/nodecves/suppress"
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+
+	api := NewAPI(defaultConf())
+	grpcServiceHandler := &supressCveServiceTestErrorImpl{}
+	api.Register(grpcServiceHandler)
+	a.Assert().NoError(api.Start().Wait())
+	a.T().Cleanup(func() { api.Stop() })
+
+	b := strings.NewReader(`{"cves": ["ABC", "XYZ"],  "duration": "24h"}`)
+
+	req, err := http.NewRequest(http.MethodPatch, url, b)
+	a.NoError(err)
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	a.NoError(err)
```

Comment:
Why is this `NoError` when `SuppressCVEs` mock implementation above sends an error? Is it because the client does not return an error if the response body has an error?

Reply:
Correct, http client returns and error when there is a connection issue but when server returns non 2xx code there is no error but code need to be handled.
> [An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem). A non-2xx status code doesn't cause an error.](https://pkg.go.dev/net/http#Client.Do)